### ISSUE TITLE
About Multi-task Learing

### ISSUE THREAD

**Issue by felix-wt** [aai-institute/tianshou#484]
- [ ] I have marked all applicable categories:
    + [x] exception-raising bug
    + [x] RL algorithm bug
    + [x] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [ ] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, numpy, sys
  print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```
  ```txt
  # 0.4.3 1.9.1+cu102 1.20.3 3.7.11 (default, Jul 27 2021, 14:32:16)  [GCC 7.5.0] linux
  ```
**Introduction:**
I am using Tianshou to do a task about robots. Using multi-task learning: the perception layer uses CNN to extract features, multiple tasks share features, and then connect to multiple FC layers。My RL policy is DQN.  Layer1 and layer2 both are trained by RL, and they share a buffer.
My network structure is as follows。
  ```python
class MyModel(nn.Module):
    def __init__(self):
        self.shared_layer = resnet18(False)
        self.layer1= nn.Sequential(... )
        self.layer2= nn.Sequential(... )
        self.layer3= nn.Sequential(... )

    def forward(self, x, state,info = {}):
        x = self.shared_layer(x)
        out1 = self.layer1(x)
        out2 = self.layer2(x)
        out3 = self.layer3(x)
        logits = {
            "out1 ":out1,
            "out2":out2,
            "out3":out3}
        return logits,state
  ```
**Question:**
I had to change the structure of the logits, because all three parts of the output are needed when outputting the action. Many errors will be reported in the `process_fn()` of DQN and when `calculating target_q`, because of logits is not a torch.Tensor. So I want to know, can the current version of Tianshou support such a multi-task structure? Can you give me advice?

**Comment by Trinkle23897** [comment]
> So I want to know, can the current version of Tianshou support such a multi-task structure?

Yes you are right, currently the code doesn't support multi-task structure. But it is still easy to support this feature:

0. create MultiTaskDQN that inherits from DQNPolicy
1. in `MultitaskDQN._target_q`, add an extra argument `key`: 
```python
    def _target_q(self, buffer: ReplayBuffer, indices: np.ndarray, key: str) -> torch.Tensor:
        batch = buffer[indices]  # batch.obs_next: s_{t+n}
        result = self(batch, input="obs_next")
        if self._target:
            # target_Q = Q_old(s_, argmax(Q_new(s_, *)))
            target_q = self(batch, model="model_old", input="obs_next").logits[key]
        else:
            target_q = result.logits[key]
        if self._is_double:
            return target_q[np.arange(len(result.act)), result.act[key]]
        else:  # Nature DQN, over estimate
            return target_q.max(dim=1)[0]
```
2. in `process_fn`:
```python
    def process_fn(self, batch: Batch, buffer: ReplayBuffer, indices: np.ndarray) -> Batch:
        returns, weight = {}, {}
        for key in ["out1", "out2", "out3"]:
            target_q_fn = lambda buffer, indices: self._target_q(buffer, indices, key)
            batch = self.compute_nstep_return(
                batch, buffer, indices, target_q_fn, self._gamma, self._n_step,
                self._rew_norm
            )
            returns[key] = batch.returns
            weight[key] = batch.weight
        for key in ["out1", "out2", "out3"]:
            batch.returns = Batch(returns)
            batch.wright = Batch(weight)
        return batch
```
3. customize the `learn` function to specify how to calculate loss

**Comment by felix-wt** [comment]
> Yes you are right, currently the code doesn't support multi-task structure. But it is still easy to support this feature:

Thanks your reply, I have solved this problem.
